File: benchmark_controls_release/public/case_000025_clean/src/app/services/audit_service.py
"""
Audit logging service — records user actions for compliance and operational review.
"""
import sqlite3
from flask import request, g, current_app


def log_action(username: str, action: str, details: str = ""):
    """Write an entry to the audit log. Non-fatal — errors are swallowed."""
    try:
        ip_addr = request.remote_addr if request else "internal"
        db_path = current_app.config["DB_PATH"]
        conn = sqlite3.connect(db_path)
        conn.execute(
            "INSERT INTO audit_log (username, action, details, ip_addr) VALUES (?,?,?,?)",
            (username, action, details[:500], ip_addr)
        )
        conn.commit()
        conn.close()
    except Exception as exc:
        current_app.logger.warning("audit_log write failed: %s", exc)

File: benchmark_release/public/case_000028/src/app/services/auditService.js
'use strict';

const { getDb } = require('../db');

/**
 * Record an auditable action taken by a user.
 */
function record({ actorId, actorName, action, resource, detail, ipAddr }) {
  try {
    const db = getDb();
    db.prepare(`
      INSERT INTO audit_log (actor_id, actor_name, action, resource, detail, ip_addr)
      VALUES (?, ?, ?, ?, ?, ?)
    `).run(actorId || 0, actorName || 'system', action, resource || '', detail || '', ipAddr || '');
  } catch (err) {
    // Non-fatal — log and continue
    require('./logger').error(`Audit record failed: ${err.message}`);
  }
}

module.exports = { record };